### PR TITLE
Updated pedigree sim and tests in C

### PR DESCRIPTION
**Pull request by nikbaya**
Updated WF pedigree model to match the Python implementation.

Modified existing tests for the pedigree simulations.

### CODE REVIEW

File: lib/msprime.c [hunk 1]
```diff
@@ -2116,7 +2126,8 @@ alloc_individual(individual_t *ind, size_t ploidy)
     // Better to allocate these as a block?
     ind->parents = malloc(ploidy * sizeof(individual_t *));
     ind->segments = malloc(ploidy * sizeof(avl_tree_t));
-    if (ind->parents == NULL || ind->segments == NULL) {
+    ind->nodes = malloc(ploidy * sizeof(avl_node_t *));
```
Comment (jeromekelleher) [inline lib/msprime.c:2129]: This is the wrong size - should be ``sizeof(tsk_id_t)`` or (better) ``sizeof(*ind->nodes)``, so you don't have to write down what the type is twice.

File: lib/msprime.c [hunk 2]
```diff
@@ -3169,15 +3140,14 @@ msp_merge_ancestors(msp_t *self, avl_tree_t *Q, population_id_t population_id,
                 }
             }
         } else {
-            if (!coalescence) {
+            if (new_node_id==TSK_NULL) {
                 coalescence = true;
                 l_min = l;
-                ret = msp_store_node(self, 0, self->time, population_id, individual);
+                new_node_id = (tsk_id_t) msp_store_node(self, 0, self->time, population_id, individual);
```
Comment (jeromekelleher) [inline lib/msprime.c:3146]: Need to check the error here:
```c
if (new_node_id < 0) {
    ret = (int) new_node_id;
    goto out;
}
```

Comment (jeromekelleher) [reply lib/msprime.c:3146]: I think you also need to modify msp_store_node to return the node ID.

File: lib/msprime.c [hunk 3]
```diff
@@ -4366,55 +4340,72 @@ msp_pedigree_climb(msp_t *self)
...
                 }
-                continue;
-            }
-
-            /* Recombine and climb to segments to the parents */
-            if (rate_map_get_total_mass(&self->recomb_map) > 0) {
-                ret = msp_dtwf_recombine(self, merged_segment, &u[0], &u[1]);
-                if (ret != 0) {
-                    goto out;
+                x = merged_head;
```
Comment (jeromekelleher) [inline lib/msprime.c:4365]: We can do this a bit more concisely in C:
```c
for (x = merged_head; x != NULL; x = x->next) {
    ...
}
```

File: lib/msprime.c [hunk 4]
```diff
@@ -4366,55 +4340,72 @@ msp_pedigree_climb(msp_t *self)
...
+            ret = msp_merge_ancestors(self, segments, 0, 0, &merged_head, parent_id, node);
             if (ret != 0) {
                 goto out;
             }
-            if (merged_segment == NULL) {
+            if (merged_head == NULL) {
                 // This lineage has coalesced
                 continue;
             }
+            ret = msp_flush_edges(self);
```
Comment (jeromekelleher) [inline lib/msprime.c:4351]: I think we should only have the flush edges at the end of the loop - we're doing it twice here, right?

File: lib/msprime.c [hunk 5]
```diff
@@ -7103,6 +7104,7 @@ msp_set_simulation_model_wf_ped(msp_t *self)
     }
 
     parents = malloc(num_individuals * self->ploidy * sizeof(*parents));
+    nodes = malloc(num_individuals * self->ploidy * sizeof(*nodes));
```
Comment (jeromekelleher) [inline lib/msprime.c:7107]: We don't need the parents, nodes, etc arrays at all now, right?

**Review by jeromekelleher — COMMENTED** [review]
Looks good!

It might be a good idea to make the change msp_merge_ancestors first separately though, and to make sure this all works by running it through CI. Always good to do things bit-by-bit and try to break as few things as possible at one time.

Could be simplest to just do it as a separate PR.

File: lib/msprime.c (hunk 2 above)
Comment (nikbaya) [reply lib/msprime.c:3146]: I think tsk_node_table_add_row already returns the node ID to msp_store_node (see the return of tsk_node_table_add_row_internal)

**Comment by nikbaya** [comment]
> It might be a good idea to make the change msp_merge_ancestors first separately though, and to make sure this all works by running it through CI. Always good to do things bit-by-bit and try to break as few things as possible at one time.

@jeromekelleher Argh yeah, I totally forgot that changing msp_merge_ancestors would break things outside of the pedigree sims because I was just running the pedigree test suite. Starting a new branch now!

**Comment by jeromekelleher** [comment]
> @jeromekelleher Argh yeah, I totally forgot that changing msp_merge_ancestors would break things outside of the pedigree sims because I was just running the pedigree test suite. Starting a new branch now!

Haha, compile and run tests early and often!

**Comment by jeromekelleher** [comment]
I think we can close this @nikbaya?

**Pull request closed without merging**